controllers/controlplaneupgrade_controller.go

```diff
@@ -223,8 +228,21 @@ func (r *ControlPlaneUpgradeReconciler) updateStatus(ctx context.Context, log lo
 			if err != nil {
 				return err
 			}
-			log.Info("Updating K8s version in  machine", "Machine", machine.Name)
+			log.Info("Updating K8s version and kubeadmClusterConfiguration annotation in machine", "Machine", machine.Name)
 			machine.Spec.Version = &nodeUpgrade.Spec.KubernetesVersion
+			kcpSpec := &controlplanev1.KubeadmControlPlaneSpec{}
+			decodedKcpSpec, err := base64.StdEncoding.DecodeString(cpUpgrade.Spec.ControlPlaneSpecData)
+			if err != nil {
+				return fmt.Errorf("decoding cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+			}
+			if err := json.Unmarshal(decodedKcpSpec, kcpSpec); err != nil {
+				return fmt.Errorf("unmarshaling cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+			}
+			clusterConfig, err := json.Marshal(kcpSpec.KubeadmConfigSpec.ClusterConfiguration)
+			if err != nil {
+				return fmt.Errorf("marshaling KCP cluster configuration: %v", err)
+			}
```

Comment:
Nit: Can we move this all to a separate function that just returns either the err or the string?